### ISSUE TITLE
Overview of Special folder names for Classes

### ISSUE THREAD

**Issue by adcxyz** [supercollider/supercollider#4968]
## Motivation
SuperCollider runs on many platforms, with different editors. Some sets of eatures (Qt, IDE, ...) are only present on some of them. When writing quarks, it is difficult to know where to put  feature-dependent classes, such that they get included or ignored correctly on all platforms.  

## Description of Proposed Feature
We should have a table in SC doc that lists which folder names get included selectively, 
both in SCClassLibrary itself and in all Quarks. 

## Plan for Implementation
We start a basic lookup table here, and invite the users on the more exotic platforms 
to contribute their knowledge here.

### Special folder names in SuperCollider 
- which folders are included/excluded in which platform/editor combinations?

### editor-specific 
editor | SCIDE | emacs | vim | Atom | headless 
------ | ------ | ------ | ---- | ------- | ----
GUI		| +	| +	| +	| +	|  -
scide_scqt | +	|-	| -	| -	|  ?
others? | ------ | ------ | ---- | ------- | ----

#### OS-specific 
Folder | macOS | linux | win | raspi | Bela | BBB
------ | -------- | ---- | ----  |-------|----|---
osx	   | +	  | -	    | -	| -	|  - | -
linux    | -	  | +	    | -	| ?	|  ? | ? | ?
win	   | - 	  | -	    | +	| -	|  - | -
others? | ------ | ------ | ---- | ------- | ----

**Comment by bgola** [comment]
Hi @adcxyz , I would add an entry for non-editor (simply running sclang).

From my experience:

- scide_scqt is only included when running from SCIDE
- GUI folder **is included** even when SC is compiled without support to X11 / SCIDE or any other graphical editor.

**Comment by redFrik** [comment]
no, one can run sclang without scide and still use the qt classes.  (like launching sclang from a script, boot the server and build a gui in the startup.scd file - it's really handy!).  what matters is if sc is **built** with qt or not.  this is what the magic scide_scqt folder is for - hiding classes from the sc compiler so that it does not call any qt primitives in initClass.
(afaiu)

**Comment by bgola** [comment]
@redFrik I also thought that, but I did a simple test here (archlinux, SC 3.11) and it doesn't work like that:

```
$ cat ~/.local/share/SuperCollider/Extensions/ARandomClass/scide_scqt/GUI/ARandomClass.sc
ARandomClass {
}
$ sclang
compiling class library...
        Found 862 primitives.
        Compiling directory '/usr/share/SuperCollider/SCClassLibrary'
        Compiling directory '/usr/share/SuperCollider/Extensions'
        Compiling directory '/home/bgola/.local/share/SuperCollider/Extensions'
        Compiling directory '/home/bgola/.local/share/SuperCollider/downloaded-quarks/adclib'
        Compiling directory '/home/bgola/.local/share/SuperCollider/downloaded-quarks/JITLibExtensions'
        Compiling directory '/home/bgola/.local/share/SuperCollider/downloaded-quarks/Influx'
        numentries = 1689222 / 35658792 = 0.047
        8262 method selectors, 4316 classes
        method table size 30503616 bytes, big table size 285270336
        Number of Symbols 26602
        Byte Code Size 2182091
        compiled 993 files in 0.75 seconds

Info: 17 methods are currently overwritten by extensions. To see which, execute:
MethodOverride.printAll

compile done
localhost : setting clientID to 0.
internal : setting clientID to 0.
Class tree inited in 0.04 seconds

*** Welcome to SuperCollider 3.11.0. *** For help type ctrl-c ctrl-h (Emacs) or :SChelp (vim) or ctrl-U (sced/gedit).
sc3> ARandomClass
ERROR: Class not defined.
  in interpreted text
  line 1 char 12:

  ARandomClass

-----------------------------------
-> nil
```

While in SCIDE the class exists.

**Comment by redFrik** [comment]
thank you @bgola.  you are right!  i stand corrected and am now very confused.

btw, scide_4 also work!  so scide_scqt could be anything starting with scide_ plus one or more characters.

**Comment by snappizz** [comment]
Would be great if all info could be compiled into this page, which is meant to exhaustively describe the patch searching mechanisms used in SC: https://github.com/supercollider/supercollider/wiki/Path-searching

**Comment by redFrik** [comment]
ok, maybe i'm not so confused after all.  i _knew_ i had used sclang without scide *and* included/excluded folders with this naming trick.

here's how i sometimes set up supercollider to start without scide...
```sInstruments.app/Contents/MacOS/sclang -a -i scqt -l sclang_conf.yaml code/loader.scd```

i double checked now and it's the `-i` flag that does it!  if i leave it out the magic folders will be ignored - like in @bgola's example above.

so sclang isn't checking if it's compiled with qt or not, but rather uses that flag to determine if scide_scqt folder should be ignored or not.
it could also be that the `-a` flag (standalone) is involved here.  did not check that.

**Comment by brianlheim** [comment]
this is the function that determines whether a directory is skipped during compilation: https://github.com/supercollider/supercollider/blob/develop/lang/LangSource/PyrLexer.cpp#L1974

most of the work is done here: https://github.com/supercollider/supercollider/blob/develop/common/SC_Filesystem.hpp#L144-L153

that should tell you pretty much everything you need to know. :)

"GUI" is never an excluded folder, it is not even a 'special name'. we just [don't install those files on builds without Qt](https://github.com/supercollider/supercollider/blob/develop/CMakeLists.txt#L455).

bela, rpi, bbb are not operating systems. they typically run Linux, so you can consider them grouped under "linux" in the table. there are no other OS's we support. nominally iPhone, but in actuality SC doesn't build there.

> We should have a table in SC doc that lists which folder names get included selectively,
both in SCClassLibrary itself and in all Quarks.

there is no difference whether files are being included from SCClassLibrary or from Quarks.

can you perhaps be more specific than 'a table in SC doc'? do you want to add this to an existing document, or in an entirely new document? as far as i can tell, we don't seem to have an article about writing or contributing quarks, which is the first place i would think to put it.

**Comment by bgola** [comment]
> that should tell you pretty much everything you need to know. :)
> 
> "GUI" is never an excluded folder, it is not even a 'special name'. we just don't install those files on builds without Qt.

So I would say there should be some mechanism to avoid compiling GUI-related classes. I see there is the -i for the IDE flag, but if I'm running a setup without graphical interface and wanna use some Quark that have classes referring to GUI stuff but not specific to any IDE it might fail to compile (because the GUI/ classes are not installed if compiled without QT for example)

**Comment by jamshark70** [comment]
> "GUI" is never an excluded folder, it is not even a 'special name'. we just don't install those files on builds without Qt.

That's a legitimate architectural weakness: Extensions are not subject to the build scripts' install logic. So, currently, in fact, there is no way to publish a quark that includes GUI extensions when run in any Qt-enabled build but excludes them in any non-Qt build -- because "Qt on or off" as a build feature is independent of the editor being used (including command line). So tying folder exclusion to the editor, and only the editor, doesn't satisfy all the requirements.

**Comment by brianlheim** [comment]
@bgola @jamshark70 i agree with you, but perhaps you could open another ticket about it? this ticket is just about documentation.

**Comment by adcxyz** [comment]
OK, so that ticket should be something like:
sclang needs separate boot flags for editor/headless and Qt/GUI,
and a short summary of the discussion here as motivation, yes?

**Comment by adcxyz** [comment]
> We have "Writing Classes", that might be a place to be considered.

Probably the best existing place would be "Using Quarks", in the section "Making your own quarks". This is where contributors are most likely to look, and should really find info on how to add platform/editor/nonGui-specific classes so they are handled properly everywhere.

Short pointers and links to it could be in multiple likely search places:
In  "Using Quarks", right after the section "Installing Quarks" as troubleshooting hints, 
definitely also in "Writing Classes", and Quarks.help. 

@brianlheim: thanks for pointing out the source file! I tried search for it with "scide_scqt", 
which did not find anything, because the code actually only uses "scide" and ignores the scqt part ;-)

**Comment by adcxyz** [comment]
started on the doc issue here: 
https://github.com/adcxyz/supercollider/tree/documentPlatformSpecificFolders

**Comment by brianlheim** [comment]
if there is a section in "Using Quarks" about writing your own, it could go there, although i would definitely not think to look for it there. we have "writing help", "writing classes", etc., and "using quarks" sounds like it is a guide for, well, using them rather than writing them.

thanks for starting on it! looks good, i left a comment where you had question marks.

i don't have any opinion about the new issue, so long as it explains the problem clearly.

**Comment by adcxyz** [comment]
@brianlheim wrote
> thanks for starting on it! looks good, i left a comment where you had question marks.

thanks! I don't see where  the comments are exactly?

**Comment by brianlheim** [comment]
oh, hm i guess it never got posted, maybe i closed the window too early. i just said you can check the editor subprojects to see how they invoke sclang to fill in the rest of the table.
